File: src/main/java/net/devfit/petclinic/controller/OwnersController.java
package net.devfit.petclinic.controller;

import java.util.List;

import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Controller;

import lombok.RequiredArgsConstructor;
import net.devfit.petclinic.controller.api.OwnersControllerApi;
import net.devfit.petclinic.domain.model.Owner;
import net.devfit.petclinic.domain.service.OwnersService;

@Controller
@RequiredArgsConstructor
public class OwnersController implements OwnersControllerApi {

  private final OwnersService ownersService;

  @Override
  public ResponseEntity<Owner> createOwner(Owner owner) {
    return null;
  }

  @Override
  public ResponseEntity<Void> deleteOwner(Integer ownerId) {
    return null;
  }

  @Override
  public ResponseEntity<List<Owner>> queryOwners(String lastName) {
    return null;
  }

  @Override
  public ResponseEntity<Owner> showOwner(Integer ownerId) {
    return ResponseEntity.ok(ownersService.showOwner(ownerId));
  }

  @Override
  public ResponseEntity<Owner> updateOwner(Integer ownerId, Owner owner) {
    return null;
  }
}

File: src/main/kotlin/net/be_geek/petclinic/controller/OwnersController.kt
package net.be_geek.petclinic.controller

import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.media.Content
import io.swagger.v3.oas.annotations.responses.ApiResponse
import io.swagger.v3.oas.annotations.responses.ApiResponses
import net.be_geek.petclinic.dto.Owner
import net.be_geek.petclinic.service.OwnersService
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.web.bind.annotation.*
import org.springframework.web.server.ResponseStatusException

@RestController
@RequestMapping("/api/owners", produces = [MediaType.APPLICATION_JSON_VALUE])
class OwnersController(val ownersService: OwnersService) {

  @Operation(summary = "Finds Owners by LastName", operationId = "queryOwners")
  @GetMapping
  fun query(@RequestParam(value = "lastName", required = false) lastName: String?): List<Owner> {
    return if (lastName == null || lastName.isEmpty()) {
      ownersService.findAll()
    } else {
      ownersService.findAllByLastName(lastName)
    }
  }

  @Operation(summary = "Find Owner by id", operationId = "showOwner")
  @ApiResponses(value = [
    ApiResponse(responseCode = "200", description = "OK"),
    ApiResponse(responseCode = "404", description = "Owner not found", content = [ Content() ])
  ])
  @GetMapping("/{ownerId}")
  fun show(@PathVariable("ownerId") ownerId: Int): Owner {
    return ownersService.findById(ownerId)
        .orElseThrow { ResponseStatusException(HttpStatus.NOT_FOUND, "Owner not found") }
  }

  @Operation(summary = "Create Owner", operationId = "createOwner")
  @PostMapping
  fun create(@RequestBody owner: Owner): Owner {
    return ownersService.create(owner)
  }

  @Operation(summary = "Update Owner", operationId = "updateOwner")
  @PutMapping("/{ownerId}")
  fun update(@PathVariable("ownerId") ownerId: Int,
             @RequestBody owner: Owner): Owner {
    return ownersService.update(ownerId, owner)
  }

  @Operation(summary = "Delete Owner", operationId = "deleteOwner")
  @DeleteMapping("/{ownerId}")
  fun delete(@PathVariable("ownerId") ownerId: Int) {
    ownersService.delete(ownerId)
  }

}
